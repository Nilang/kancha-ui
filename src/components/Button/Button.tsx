import * as React from 'react'
import { TouchableHighlight, TouchableOpacity, ViewStyle } from 'react-native'
import { withTheme } from '../../theming'
import * as Kancha from '../../types'
import Text, { TextTypes } from '../Text/Text'
import Container from '../Container/Container'

export const ButtonBlocks: Kancha.BlocksStatic = {
  Outlined: 'outlined',
  Filled: 'filled',
  Clear: 'clear',
}

interface ButtonProps {
  /**
   * The button type. This sets the theme color
   */
  type?: Kancha.BrandPropOptions
  /**
   * The block appearance of the button
   */
  block?: Kancha.BlockPropsOptions
  /**
   * The text to be displayed
   */
  buttonText?: string

  /**
   * The text to be displayed
   */
  bold?: boolean

  /**
   * Remove width limitations
   */
  fullWidth?: boolean
  /**
   * The button type. This sets the theme color
   */
  onPress: () => void

  /**
   * Center the button horizontally on screen
   */
  centered?: boolean

  /**
   * Disbable the button
   */
  disabled?: boolean

  /**
   * The button is a navigation button
   */
  navButton?: boolean

  /**
   * Provide an icon component
   */
  icon?: React.ReactNode

  /**
   * Remove button padding for icon buttons
   */
  noPadding?: boolean

  /**
   * Text decoration
   */
  textDecorationLine?: 'none' | 'underline' | 'line-through' | 'underline line-through' | undefined

  /**
   * Provide a testID for e2e tests
   */
  testID?: string

  /**
   * Shadow depth
   */
  depth?: number

  /**
   * Shadow depth
   */
  iconButton?: boolean

  /**
   * Theme used by theme provider
   */
  theme: any
}

const Button: React.FC<ButtonProps> & {
  Block: Kancha.BlocksStatic
} = ({
  type,
  block,
  fullWidth,
  onPress,
  disabled,
  buttonText,
  centered,
  bold,
  navButton,
  icon,
  noPadding,
  textDecorationLine,
  testID,
  iconButton,
  theme,
}) => {
  const style: ViewStyle = {
    ...(block && block === 'filled'
      ? { backgroundColor: type ? theme.colors[type].button : theme.colors.primary.button }
      : {}),
    ...(block && block === 'outlined'
      ? {
          backgroundColor: theme.colors.primary.background,
          borderWidth: 2,
          borderColor: type ? theme.colors[type].button : theme.colors.primary.button,
        }
      : {}),
    ...(noPadding || iconButton ? {} : { padding: theme.spacing.default }),
    alignItems: 'center',
    ...(fullWidth ? {} : { maxWidth: 300 }),
    borderRadius: theme.roundedCorners.buttons,
    ...(centered ? { alignSelf: 'center' } : {}),
    ...(disabled ? { opacity: 0.2 } : {}),
  }

  return navButton ? (
    <TouchableOpacity
      style={style}
      onPress={onPress}
      disabled={disabled}
      testID={testID}
      accessibilityLabel={testID}
    >
      <Text
        textDecorationLine={textDecorationLine}
        type={TextTypes.NavButton}
        buttonTextColor={disabled ? 'secondary' : type}
        block={block}
      >
        {buttonText}
      </Text>
    </TouchableOpacity>
  ) : (
    <TouchableHighlight
      testID={testID}
      accessibilityLabel={testID}
      disabled={disabled}
      onPress={onPress}
      style={style}
      underlayColor={
        block === ButtonBlocks.Clear || iconButton ? 'transparent' : type && theme.colors[type].underlay
      }
    >
      <Container flexDirection={'row'}>
        {icon && icon}
        <Text
          textDecorationLine={textDecorationLine}
          type={TextTypes.Button}
          buttonTextColor={disabled ? 'secondary' : type}
          block={block}
          bold={bold}
        >
          {buttonText}
        </Text>
      </Container>
    </TouchableHighlight>
  )
}

Button.Block = ButtonBlocks

export default withTheme(Button)
