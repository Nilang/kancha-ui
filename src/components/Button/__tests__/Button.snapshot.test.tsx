import React from 'react'
import { render } from 'react-native-testing-library'

import Button from '../Button'

describe('Container', () => {
  it('renders correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a nav button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} navButton={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('renders a disabled button correctly', () => {
    const onPress = jest.fn()
    const tree = render(<Button onPress={onPress} disabled={true} />).toJSON()

    expect(tree).toMatchSnapshot()
  })
})
