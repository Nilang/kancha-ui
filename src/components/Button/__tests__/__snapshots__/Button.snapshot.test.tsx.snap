// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshots): Button renders a disabled button correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "opacity": 0.2,
      "padding": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
        }
      }
    />
  </View>
</View>
`;

exports[`Component(snapshots): Button renders a disabled nav button correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "opacity": 0.2,
      "padding": 15,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#FFFFFF",
        "fontSize": 20,
      }
    }
  />
</View>
`;

exports[`Component(snapshots): Button renders a filled button correctly with all props 1`] = `
<View
  accessibilityLabel="TEST_ID"
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#6A54D1",
      "borderRadius": 8,
    }
  }
  testID="TEST_ID"
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    />
  </View>
</View>
`;

exports[`Component(snapshots): Button renders a nav button correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
      "opacity": 1,
      "padding": 15,
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 20,
      }
    }
  />
</View>
`;

exports[`Component(snapshots): Button renders an icon button correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderRadius": 8,
      "maxWidth": 300,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    />
  </View>
</View>
`;

exports[`Component(snapshots): Button renders an outlined button correctly 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#6A54D1",
      "borderRadius": 8,
      "borderWidth": 2,
      "maxWidth": 300,
      "padding": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
        }
      }
    />
  </View>
</View>
`;

exports[`Component(snapshots): Button renders an outlined button correctly with a type 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "borderColor": "#6A54D1",
      "borderRadius": 8,
      "borderWidth": 2,
      "maxWidth": 300,
      "padding": 15,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "flex": undefined,
        "flexDirection": "row",
        "height": undefined,
        "justifyContent": undefined,
        "marginBottom": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "width": undefined,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#6A54D1",
          "fontSize": 18,
        }
      }
    />
  </View>
</View>
`;
